rfcs/0028-dag-structured_query_plan.md

```diff
@@ -0,0 +1,67 @@
+
+## Design
+
+The DAG structured query plan can be supported by optimizer and executor independently. If the executor doesn't support DAG, we can always transform the DAG plan back to a tree plan. For example, if our optimizer can generate a DAG plan, but only streaming executor can support DAG execution and batch executor can't. We can convert it back to a tree plan before we send it to the batch executor. In reality I think streaming executor can benefit from DAG plan more than batch executor.
+
+### Optimizer
+
+The simplest way to implement a DAG plan for the optimizer is that we can place an optimization stage at the end of all other stages and try to find out all the common sub-plan and transform it to a DAG plan. It is simple and also adopted by flink. One of the biggest drawbacks of this method is it is hard to find any sub-plans after the prior optimization: column pruning and predicate push down etc.
+
+### Share Operator
+
+In order to represent the DAG plan during the whole optimization time, we can use a `Share` operator to explicitly represent the sub-plan having multiple parents. Why do we need to add an additional operator instead of directly letting its parent point to the sub-plan? Because we have lots of optimizer transformations which will clone and modify the subquery and finally break the DAG plan back to a tree plan. Using a `Share` operator we can tell the optimizer we want to keep the sub-plan unchanged during the optimization.
+
+![](images/share_operator.svg)
+
+The `Share` operator contains multiple parents. Each of them wants to push the predicate and prune the columns of it, so we need to merge them instead of directly affecting the sub-plan. We can create a new optimization stage for merging all the predicates pushed down from the parent of the `Share` operator. The merge operation for filters is to `OR` them together. We can merge the required columns from the parents in a similar way. The merge operation for the column pruning is to `UNION` them together.
+
+
+#### Example
+With the `Share` operator we can perform common plan detection at any time of the optimization. I will give some examples:
+
+- CTE. At the binder/planner stage, we can create a `Share` for the table expression and reuse it.
+- Subquery unnesting. We can calculate the domain by creating a `Share` operator of the LHS of the apply operator.
+- Same source operator appears multiple times in a query. Use Share operator at binder/planner or at end of optimizer are both ok.
+- Common sub-plan detection. We can use a field `digest` for each sub-plan and if two sub-plan have the same `digest` it means they are probably the same sub-plan. `digest` is just like a hashcode. We also need to have a `deepEqual` method for each plan to compare with another one.
+
+![](images/self_join.svg)
```

Comment:
FYI, executing a graph with multiple-edges is currently not a simple deal in the compute node. 🥵
https://www.notion.so/risingwave-labs/Multiple-Edges-in-the-Stream-Graph-f0ab595523d74823a39edaebdea2be16